File: ios/Sources/Features/Shared/LocalDataFootprint.swift
import Foundation

/// On-device Application Support footprints used by Settings.
enum LocalDataFootprint {
    static func knowledgeBaseByteCount(fileManager: FileManager = .default) -> Int64 {
        directoryByteCount(named: "KnowledgeBase", fileManager: fileManager)
    }

    static func chatHistoryByteCount(fileManager: FileManager = .default) -> Int64 {
        directoryByteCount(named: "KnowledgeChat", fileManager: fileManager)
    }

    static func formattedByteCount(_ bytes: Int64) -> String {
        ByteCountFormatter.string(fromByteCount: max(0, bytes), countStyle: .file)
    }

    private static func directoryByteCount(
        named name: String,
        fileManager: FileManager
    ) -> Int64 {
        guard let root = try? fileManager.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: false
        ) else { return 0 }
        let url = root.appendingPathComponent(name, isDirectory: true)
        return recursiveByteCount(at: url, fileManager: fileManager)
    }

    private static func recursiveByteCount(at url: URL, fileManager: FileManager) -> Int64 {
        var isDirectory: ObjCBool = false
        guard fileManager.fileExists(atPath: url.path, isDirectory: &isDirectory) else { return 0 }
        if !isDirectory.boolValue {
            let values = try? url.resourceValues(forKeys: [.fileSizeKey])
            return Int64(values?.fileSize ?? 0)
        }
        guard let enumerator = fileManager.enumerator(
            at: url,
            includingPropertiesForKeys: [.isRegularFileKey, .fileSizeKey],
            options: [.skipsHiddenFiles]
        ) else { return 0 }
        var total: Int64 = 0
        for case let fileURL as URL in enumerator {
            guard let values = try? fileURL.resourceValues(forKeys: [.isRegularFileKey, .fileSizeKey]),
                  values.isRegularFile == true else { continue }
            total += Int64(values.fileSize ?? 0)
        }
        return total
    }
}

File: android/app/src/main/kotlin/ai/orynode/mobile/app/ui/LocalDataFootprint.kt
package ai.orynode.mobile.app.ui

import android.content.Context
import java.io.File
import java.util.Locale

/** On-device footprints used by Settings (parity with iOS LocalDataFootprint). */
object LocalDataFootprint {
    fun knowledgeBaseByteCount(context: Context): Long =
        directoryByteCount(File(context.filesDir, "KnowledgeBase"))

    fun chatHistoryByteCount(context: Context): Long =
        directoryByteCount(File(context.filesDir, "KnowledgeChat"))

    fun formattedByteCount(bytes: Long): String {
        val value = maxOf(0L, bytes).toDouble()
        return when {
            value >= 1_000_000_000 -> String.format(Locale.US, "%.2f GB", value / 1_000_000_000)
            value >= 1_000_000 -> String.format(Locale.US, "%.1f MB", value / 1_000_000)
            value >= 1_000 -> String.format(Locale.US, "%.0f KB", value / 1_000)
            else -> "$bytes B"
        }
    }

    private fun directoryByteCount(root: File): Long {
        if (!root.exists()) return 0
        if (root.isFile) return root.length()
        var total = 0L
        root.walkTopDown().forEach { file ->
            if (file.isFile) total += file.length()
        }
        return total
    }
}
